ipld/get_shares.go

```diff
@@ -0,0 +1,101 @@
+package ipld
+
+import (
+	"context"
+	"sync"
+
+	"github.com/gammazero/workerpool"
+	"github.com/ipfs/go-cid"
+	format "github.com/ipfs/go-ipld-format"
+)
+
+// NumWorkersLimit sets global limit for workers spawned by GetShares.
+// GetShares could be called MaxSquareSize(128) times per data square each spawning up to 128/2 goroutines and
+// altogether this is 8192. Considering there can be N blocks fetched at the same time, e.g. during catching up data
+// from the past, we multiply this number by the amount of allowed concurrent data square fetches(NumConcurrentSquares).
+// NOTE: This value only limits amount of simultaneously running workers that are spawned as the load increases and
+// are killed, once the load declines. Thanks to simple yet powerful workerpool library by @gammazero.
+// TODO(@Wondertan): This assumes we have parallelized DASer implemented. Sync the values once it is shipped.
+// TODO(@Wondertan): Allow configuration of values.
+var NumWorkersLimit = MaxSquareSize * MaxSquareSize / 2 * NumConcurrentSquares
```

Comment:
Is there a reason not to declare these as constants?

Reply:
So we at least have some control over those. Specifically, I played around with those in reconstruction tests.
`ipld.NumWorkersLimit=<num>`

Reply:
Seems like a leaky abstraction. Is there a better way to do that while guaranteeing that the values won't be changed accidentally by the code?

Reply:
I also do not think it is great to have this exported and mutable.

Reply:
This comment is a nit, fixing it is time sink. It is value is super small comparting to what this PR tries to achieve. Pls better look at the implementaion and find issues with it. I also agree that the should not be a global var thing and this is why I made a TODO there before this converstation is started. I can change this now, but it would take more time than you might thing. The possible damage of this var being misused is negligible.

Reply:
Ok